purchase_line_service_qty_received/models/purchase_order_line.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2024 Raumschmiede GmbH
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl)
+
+from odoo import fields, models
+from odoo.tools import float_is_zero, groupby
+
+METHOD_RECEPTION = "reception"
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    qty_received_method = fields.Selection(
+        selection_add=[(METHOD_RECEPTION, "Reception")]
+    )
+
+    def _is_received_method_reception(self):
+        self.ensure_one()
+
+        return (
+            self.product_id.type == "service"
+            and self.product_id.purchase_method == "receive"
+        )
+
+    def _compute_qty_received_method(self):
+        super()._compute_qty_received_method()
+
+        for line in self:
+            if line._is_received_method_reception():
+                line.qty_received_method = METHOD_RECEPTION
+
+    def _get_reception_qty_received(self, received):
+        self.ensure_one()
+
+        if received:
+            return self.product_uom_qty
+
+        return 0.0
+
+    def _compute_qty_received_reception_method(self):
+        precision = self.env["decimal.precision"].precision_get(
+            "Product Unit of Measure"
+        )
+
+        # When picking is set to done, in self are only the lines that have a stock
+        # move connected. So take all lines of self's PO to be sure to set qty_received
+        # for the service products
+        lines_grouped = groupby(self.order_id.order_line, lambda l: l.order_id)
+        for _, lines in lines_grouped:
+            lines = dict(groupby(lines, lambda l: l.qty_received_method))
+
+            if METHOD_RECEPTION not in lines:
+                continue
+
+            received = True
+            if "stock_moves" in lines:
+                # True if at least any PO line with a stock product has a qty_received
+                received = any(
+                    not float_is_zero(line.qty_received, precision)
+                    for line in lines["stock_moves"]
+                )
+
+            for line in lines[METHOD_RECEPTION]:
```

Comment:
I think the complexity can be reduced by first filtering _[O(n)]_ instead of grouping twice _[O(n²)]_. Also better to not redefine _
```suggestion
        lines_grouped = groupby(
            self.order_id.order_line.filtered(lambda line: line.qty_received_method == METHOD_RECEPTION),
            lambda l: l.order_id
        )
        for order, lines in lines_grouped:
            # True if at least any PO line with a stock product has a qty_received
            received = any(
                    line.qty_received_method == "stock_moves"
                    and not float_is_zero(line.qty_received, precision)
                    for line in order.order_line
            )
            for line in lines:
```

Reply:
I applied the changes. They are better. I tested them and found out that without the changes a PO with only service products the PO lines with a reception service product have `qty_received` automatically set to 1. With your code not.  

And your code fixes the failing tests in the other modules :D  
Thanks a lot